Add airfoil grid generator

eliasstudiert opened the pull request:
Resulting (2D) triangulations looks like this:

![NACA2412](https://user-images.githubusercontent.com/50180840/68686526-0ebb8a80-056c-11ea-8587-d404c06a9955.png)

peterrum commented:
This PR is part of @eliasstudiert's Bachelor's thesis. Hope you like the new grid generator! It is currently limited to 2D and does not use manifolds. The documentation still has to be extended...

A first feedback is welcome!

drwells commented:
This is a nice picture and definitely something worth adding to the library!

@eliasstudiert do you have any plans to extend this to use manifolds?

eliasstudiert commented:
> This is a nice picture and definitely something worth adding to the library!
> 
> @eliasstudiert do you have any plans to extend this to use manifolds?

Thanks! I didn't think about how to incorporate Manifolds yet, but I'd be thankful for any ideas or hints.

blaisb commented:
@eliasstudiert this will be very useful for us! Nice addition.
It would be very interesting if you could add manifolds, I agree with that :)!

drwells commented:
@eliasstudiert I'll have to take a closer look, but since you have some description of the geometry already embedded in he source code (so that we can generate the grid around the airfoil) it should be possible to generalize that as a manifold.

bangerth commented:
Nice work! I only know for the NACA0012 airfoil, but for that the surface is given as an analytically known function that should be relatively easy to encode in a manifold. One would then use the `TransfiniteInterpolationManifold` class for the remainder of the mesh.

@eliasstudiert -- there are a number of TODOs or unfinished sentences still left in the documentation in the .h file. If you wanted to go through those first, we can save one round of reviews and the whole process is likely going to be a bit faster than if we have to mark all of these places up. Can you go through your text again and let us know when these are addressed?

luca-heltai commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -60,6 +61,173 @@ namespace GridGenerator
+       * @param prm Parameter handler.
+       */
+      void
+      get(ParameterHandler &prm);
```
Is there a reason not to use the `add_parameter` feature of `ParameterHandler`? This would allow you to attach a parameter right at the time of the declaration, and would remove the above `get` function (simplifying the code, and removing the possibility of bugs when adding/removing parameters in future versions).

eliasstudiert replied:
> Nice work! I only know for the NACA0012 airfoil, but for that the surface is given as an analytically known function that should be relatively easy to encode in a manifold. One would then use the `TransfiniteInterpolationManifold` class for the remainder of the mesh.
> 
> @eliasstudiert -- there are a number of TODOs or unfinished sentences still left in the documentation in the .h file. If you wanted to go through those first, we can save one round of reviews and the whole process is likely going to be a bit faster than if we have to mark all of these places up. Can you go through your text again and let us know when these are addressed?

I added the documentation and finished the TODOs. Also I simplified the function interpolate() in taking advantage of symmetry of the coarse grid. The code is ready to review. I didn't figure out so far, why `create_triangulation()` without `AdditionalData data` as inputparameter (default setup) doesn't work without providing the inputfile.

luca-heltai commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1175 @@
+        prm.declare_entry("Height",
+                          std::to_string(height),
+                          Patterns::Double(),
+                          "CoarseGridBlockHight");
```
For example, this would become:

`prm.add_entry("Height", height, "CoarseGridBlockHight");`

and you could remove the corresponding entry in the `get` function.

eliasstudiert replied:
I removed function get() in AdditionalData as you suggested and used add_parameter() to attach the parameters instead. 
Somehow, the function `create_triangulation(tria)` and `create_triangulation(tria, periodic_faces)` which should produce a default mesh since no `AdditionalData data` object is passed as input, doesn't work without the inputfile (parse_input(inputfile)).

Is it possible that this change (using add_paramter instead declare_entry() and get()) caused that issue?

luca-heltai replied:
This is totally unrelated. My guess is that the default object you create is invalid (i.e., maybe you forgot to initialize some of the members of AdditionalData), and they are initialized by the add_parameters method instead of being initialized at construction time. 

I'd move the initialization where you declare the objects, i.e.,

`
center = Point<2>(0,0);
`
instead of having them initialized in the constructor. This documents the default values in the same place where you declare the members, and ensures that no entries are left un-initialized.

The add_parameter method simply sets the values when it is called, or when a file is parsed.

eliasstudiert replied:
The mentioned problem was caused, as Peter expected, by a mistake in my test program.

Should I still move the initialization of `center` as you suggested?

peterrum replied:
@luca-heltai Initializing default values in the initializer list is how we normally do it, isn't it?

luca-heltai commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1175 @@
+        prm.declare_entry("LengthB2",
+                          std::to_string(length_b2),
+                          Patterns::Double(),
+                          "LengthBlock2");
```
same here. And the same in all the other cases.

eliasstudiert commented:
> Nice work! I only know for the NACA0012 airfoil, but for that the surface is given as an analytically known function that should be relatively easy to encode in a manifold. One would then use the `TransfiniteInterpolationManifold` class for the remainder of the mesh.
> 
> @eliasstudiert -- there are a number of TODOs or unfinished sentences still left in the documentation in the .h file. If you wanted to go through those first, we can save one round of reviews and the whole process is likely going to be a bit faster than if we have to mark all of these places up. Can you go through your text again and let us know when these are addressed?

eliasstudiert commented:
@bangerth I added the documentation and finished the TODOs. Also I simplified the function interpolate() in taking advantage of symmetry of the coarse grid. The code is ready to review. I didn't figure out so far, why create_triangulation() without AdditionalData data as inputparameter (default setup) doesn't work without providing the inputfile.

drwells commented:
@eliasstudiert I don't think you meant to close this: we can keep it open and mark as WIP until its ready :)

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -13,10 +13,12 @@
 #include <deal.II/distributed/tria.h>
 
 #include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_out.h>
```
This is not needed anymore.

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+       * I.H. Abbott and A.E. von Doenhoff. Theory of Wing Sections: Including a
+       * Summary of Airfoil Data. New York: Dover Publications, 1959.
+       *
+       * @param NACA-serialnumber for different airfoilshapes (so far only 4-digit-series implemented)
```
```suggestion
       * @param NACA-serial number for different airfoil shapes (so far only 4-digit-series implemented)
```
the same below

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+        const unsigned int digit_23 = 10 * digit_2 + digit_3;
+
+        const double t = static_cast<double>(digit_23) / 100.0;
+        // 1.0 * digit_23 / 100; // maximum thickness in percentag of the cord
```
Move comment above.

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+
+        const double t = static_cast<double>(digit_23) / 100.0;
+        // 1.0 * digit_23 / 100; // maximum thickness in percentag of the cord
+        double x, y_t;
```
You can move these into the loops below and make them const.

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+       * Set airfoil length (i.e. chord length) to a desired length.
+       * calls function set_airfoil_length() for each vector of the array
+       *
+       * @param array containing upper and lower vector
```
I just noticed. The format is:
```suggestion
       * @param input Array containing upper and lower vector.
```
Please adjust the parameters in the other functions similarly.

eliasstudiert replied:
so for each @param we use:
input <Type> <description>. 
E.g., @param input Double desired length. ?

peterrum replied:
Yes.

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+            const auto Lj  = arclength_L[j];
+            const auto Ljp = arclength_L[j + 1];
+
+
```
You can remove this line.

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+            non_equidistant_points[i + 1].distance(non_equidistant_points[i]);
+
+
+        double airfoil_length =
```
```suggestion
        const auto airfoil_length =
```

eliasstudiert replied:
how about parameters `deltaX` and `l` in the same function, can't they be constant as well?
`const double deltaX` and `const double l`

peterrum replied:
Make as much as possible `const`.

peterrum replied:
Statements at one place is true for all other places...

eliasstudiert replied:
in this function no more parameters can be const right? Sorry, but I don't understand what you mean by `Statements at one place is true for all other places...`

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+
+        double airfoil_length =
+          arclength_L.back(); // arclength upper or lower side
+        double deltaX = airfoil_length / (number_points - 1);
```
```suggestion
        const auto deltaX = airfoil_length / (number_points - 1);
```

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+                                       non_equidistant_points[j]);
+                i++;
+              }
+          }
```
```suggestion
            while (Lj <= i * deltaX && i * deltaX <= Ljp &&
                   i < number_points - 1)
                equidist[i++] = Point<2>((i * deltaX - Lj) / (Ljp - Lj) *
                                           (non_equidistant_points[j + 1] -
                                            non_equidistant_points[j]) +
                                         non_equidistant_points[j]);
          }
```

eliasstudiert replied:
warning: operation on ‘i’ may be undefined [-Wsequence-point]
                 equidist[i++] = Point<2>((i * deltaX - Lj) / (Ljp - Lj) *

Shouldn't be a problem, right? Or can we remove that warning somehow?

peterrum replied:
Warnings are bad. Does it help to move the `i++` down?

masterleinad replied:
It's not clear if `i` is first incremented and then evaluated on the right-hand side or the other way around.

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+       * Create 6 coarse grids based on points A-L (class fields) and merges
+       * them to one triangulation.
+       *
+       * @param tria
```
Just remove this line.

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+       * - 4: upper far-field side
+       * - 5: lower far-field side
+       *
+       * @param tria
```
Just remove this line.

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+       * mesh more tight in close distance of the airfoil.
+       *
+       * @param alpha
+       * @return alpha
```
Just remove these lines.

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+
+
+      /*
+       * This function returns a bias factor 'alpha' which is used to make the
```
You could mention that this is bijective function mapping from [0,1] onto [0,1]. Are values near 0 or near 1 made tighter?

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+                      rotation_matrix_1[0][0] = +std::cos(-gamma);
+                      rotation_matrix_1[0][1] = -std::sin(-gamma);
+                      rotation_matrix_1[1][0] = +std::sin(-gamma);
+                      rotation_matrix_1[1][1] = +std::cos(-gamma);
```
You should setup the matrix only once at the top.

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+                      rotation_matrix_1[0][0] = +std::cos(+gamma);
+                      rotation_matrix_1[0][1] = -std::sin(+gamma);
+                      rotation_matrix_1[1][0] = +std::sin(+gamma);
+                      rotation_matrix_1[1][1] = +std::cos(+gamma);
```
The same here. You can save it into a new `rotation_matrix_2`.

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+        Tensor<2, 2, double> rotation_matrix_1;
+
+        // horizontal offset in order to place coarse-grid node A in the origin
+        Point<2, double> horizontal_offset(A(0), 0.0);
```
```suggestion
        const Point<2, double> horizontal_offset(A(0), 0.0);
```

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+        Point<2, double> horizontal_offset(A(0), 0.0);
+
+        // Move block 1 so that face BC coincides the x-axis
+        Point<2, double> trapeze_offset(0.0, std::sin(gamma) * edge_length);
```
```suggestion
        const Point<2, double> trapeze_offset(0.0, std::sin(gamma) * edge_length);
```

eliasstudiert commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+       * Create a parallel::distributed::Triangulation.
+       */
+      void create_triangulation(
+        parallel::distributed::Triangulation<2> &parallel_grid,
```
so we can remove  function `void create_triangulation(parallel::distributed::Triangulation<2> &parallel_grid, ...)` since `void create_triangulation(Triangulation<2> &tria_grid, ...)` will work for serial and distributed triangulations, right?

peterrum replied:
I think it will work. You should try it out.

eliasstudiert commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+        mesh_generator.create_triangulation(tria, nullptr);
+      // Assert(false, ExcMessage("Mesh has been created successfully!"));
+      std::cout << "Mesh has been created successfully!" << std::endl;
+    }
```
I changed this function as you suggested, but there is a compiling error which says no matching function to call, since 

no known conversion for argument 2 from ‘std::nullptr_t’ to ‘std::vector<dealii::GridTools::PeriodicFacePair<dealii::TriaIterator<dealii::CellAccessor<2, 2> > >, std::allocator<dealii::GridTools::PeriodicFacePair<dealii::TriaIterator<dealii::CellAccessor<2, 2> > > > >&’

How could I fix this ?

peterrum replied:
You are right. The function takes a reference not a pointer. I guess you have to create an internal function `internal_create_triangulation`, which takes a pointer and is called by both `create_triangulation` functions.

eliasstudiert commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+      const double center_mesh;
+
+      // length of blocks 2 and 5
+      const double length_b1_x;
```
to be conform with `n_subdivision_x_2`. I use numbers 0 (=Blocks 1&4), 1 (=Blocks 2&5) and 2 (=Blocks 3&6). Should I use a different approach?

peterrum replied:
It is somewhat inconsistent to start one from 0 and the other from 1.

eliasstudiert commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+        std::vector<Point<2>> joukowski_points(circle_points.size());
+
+        std::complex<double> z;
+        double               chi, eta;
```
you mean, I shouldn't define parameters chi and eta at all and directly use
`const std::complex<double> zeta(circle_points[i](0), circle_points[i](1));`

Or should I move the definition 
            `const double chi = circle_points[i](0);`
            `const double eta = circle_points[i](1);`
inside the for loop ?

peterrum replied:
Yes.

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+
+        // create a subdivided quadrilateral for each block (see last number of
+        // block id)
+        make(trias[0], {A, B, D, C}, {n_subdivision_y, n_subdivision_x_0}, 1);
```
Could you make all instances of block ids static constants. You could write at the top of the class:
```cpp
static const unsigned int id_block_1 = 1;
```
and write here:
```suggestion
        make(trias[0], {A, B, D, C}, {n_subdivision_y, n_subdivision_x_0}, id_block_1);
```
This would make it easier to distinguish block ids form actual numbers.

eliasstudiert replied:
at the top of the class means right after `void make_coarse_grid(Triangulation<2> &tria) const
      {` or as class lists of MeshGenerator class?

eliasstudiert commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+            chi = circle_points[i](0);
+            eta = circle_points[i](1);
+            std::complex<double> zeta(chi, eta);
+            z = zeta + 1. / zeta;
```
You mean private variables inside a function? What is the intention of private variables inside a function?

peterrum replied:
Like I have written in code snippet.

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +30,1180 @@
+     */
+    class MeshGenerator
+    {
+    public:
```
```cpp
static const unsigned int id_block_1 = 1;
```

eliasstudiert commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -60,6 +61,179 @@ namespace GridGenerator
+     *
+     * @param tria The triangulation to be created. It needs to be empty upon
+     *             calling this function.
+     * @param additional_data Configuration of the mesh.
```
So a figure created in incscape showing a mesh schematically with the input parametes of class AdditionalData, right? I will add the figure till Friday evening.
Which format of the picture, png? And how to add a picture in the code? By the link where it is saved?

kronbichler replied:
Regarding the figure, I suggest we post it on the homepage (since it is 364 kB, so a bit large to include it in the git repo). I have uploaded the file to the homepage repo, the link is:
https://www.dealii.org/images/grids/airfoils_naca_joukowski.png

Given the resolution of the file, I suggest we insert the file with some relative width compared to screen. I'm not an expert on what markup is allowed by doxygen, but looking around in other places suggests we could do something like
```
\htmlonly <style>div.image
   * img[src="https://www.dealii.org/images/grids/airfoils_naca_joukowski.png"]{width:50%;}</style> \endhtmlonly
   * @image html https://www.dealii.org/images/grids/airfoils_naca_joukowski.png
```

peterrum commented:
@eliasstudiert Looks good now! Let's wait for some feedback from the main developers!

eliasstudiert commented:
![visualization](https://user-images.githubusercontent.com/50180840/69457746-10583000-0d6e-11ea-8a9c-4013035fa412.png)

I don't know where is the right place to add this image. It should help to understand the input parameter.

peterrum commented:
Very nice! This should help the reviewer. @bangerth can we use this picture in the docu?

blaisb commented:
> ![visualization](https://user-images.githubusercontent.com/50180840/69457746-10583000-0d6e-11ea-8a9c-4013035fa412.png)
> 
> I don't know where is the right place to add this image. It should help to understand the input parameter.

Amazing picture / amazing work. Love it

peterrum commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1189 @@
+
+      prm.enter_subsection("Joukowski");
+      {
+        prm.add_parameter("CenterX",
```
In the parameter file, you could simply write than:
```
set Center = 0, 0
```

and we can remove `CenterX` and `CenterY`.

eliasstudiert replied:
Thanks! This works.

masterleinad commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1186 @@
+      prm.enter_subsection("Joukowski");
+      {
+        prm.add_parameter("Center",
+                          center,
```
```suggestion
                          joukowski_center,
```

peterrum replied:
@masterleinad Thanks! I have changed this!

peterrum commented:
@eliasstudiert I have squashed all your commits to a single commit. What I have noticed that your name and email address in git is not set up correctly. Could you set it up as described here: https://github.com/dealii/dealii/wiki/Indentation#commit-authorship

Thanks!

kronbichler commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -60,6 +61,183 @@ namespace GridGenerator
+      AdditionalData();
+
+      /**
+       * This Function adds the ParameterHandler entries.
```
```suggestion
       * This function adds the ParameterHandler entries.
```

kronbichler commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -60,6 +61,183 @@ namespace GridGenerator
+
+    /**
+     * Initialize the given triangulation with a flow field around an airfoil,
+     * i.e. a Mesh of C-Type approximating Joukowski or NACA (4 digit) airfoils.
```
```suggestion
     * i.e., a mesh of C-Type approximating Joukowski or NACA (4 digit) airfoils.
```

kronbichler commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -60,6 +61,183 @@ namespace GridGenerator
+     * i.e. a Mesh of C-Type approximating Joukowski or NACA (4 digit) airfoils.
+     *
+     * The user can specify the airfoil geometry and the mesh setup by providing
+     * input parametrs for class AdditionalData.
```
```suggestion
     * input parameters for the struct AdditionalData.
```

kronbichler commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -60,6 +61,179 @@ namespace GridGenerator
+      double incline_factor;
+
+      /**
+       * Factor to receive a finer mesh around the airfoil by increasing it.
```
I agree that the function `tanh` should be specified.

kronbichler commented:
/rebuild

kronbichler commented:
There are errors in the MacOSX build:
```
/Users/jenkins/jenkins/workspace/dealii-OSX_PR-9045/source/grid/grid_generator.cc:177:20: error: suggest braces around initialization of subobject [-Werror,-Wmissing-braces]
                   std::vector<Point<2>>{Point<2>(0), Point<2>(1)},
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/jenkins/jenkins/workspace/dealii-OSX_PR-9045/source/grid/grid_generator.cc:566:17: error: suggest braces around initialization of subobject [-Werror,-Wmissing-braces]
        return {make_points_equidistant(
                ^~~~~~~~~~~~~~~~~~~~~~~~
/Users/jenkins/jenkins/workspace/dealii-OSX_PR-9045/source/grid/grid_generator.cc:994:39: error: implicit conversion turns floating-point number into integer: 'double' to 'const int' [-Werror,-Wfloat-conversion]
                    const int    ix = std::round((node_(0) - deltax) / dx);
                                 ~~   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/jenkins/jenkins/workspace/dealii-OSX_PR-9045/source/grid/grid_generator.cc:995:39: error: implicit conversion turns floating-point number into integer: 'double' to 'const int' [-Werror,-Wfloat-conversion]
                    const int    iy = std::round(std::abs(node_(1)) / dy);
                                 ~~   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/jenkins/jenkins/workspace/dealii-OSX_PR-9045/source/grid/grid_generator.cc:1006:42: error: implicit conversion turns floating-point number into integer: 'double' to 'const int' [-Werror,-Wfloat-conversion]
                    const int    ix    = std::round(node_(0) / dx);
                                 ~~      ^~~~~~~~~~~~~~~~~~~~~~~~~
/Users/jenkins/jenkins/workspace/dealii-OSX_PR-9045/source/grid/grid_generator.cc:1007:42: error: implicit conversion turns floating-point number into integer: 'double' to 'const int' [-Werror,-Wfloat-conversion]
                    const int    iy    = std::round(node_(1) / dy);
                                 ~~      ^~~~~~~~~~~~~~~~~~~~~~~~~
/Users/jenkins/jenkins/workspace/dealii-OSX_PR-9045/source/grid/grid_generator.cc:1038:38: error: implicit conversion turns floating-point number into integer: 'double' to 'const int' [-Werror,-Wfloat-conversion]
                  const int    iy  = std::round((std::abs(node(1)) - by) / dy);
                               ~~    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/jenkins/jenkins/workspace/dealii-OSX_PR-9045/source/grid/grid_generator.cc:1041:21: error: implicit conversion turns floating-point number into integer: 'double' to 'const int' [-Werror,-Wfloat-conversion]
                    std::round(std::abs(node(0) - center_mesh) / dx);
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/jenkins/jenkins/workspace/dealii-OSX_PR-9045/source/grid/grid_generator.cc:1067:37: error: implicit conversion turns floating-point number into integer: 'double' to 'const int' [-Werror,-Wfloat-conversion]
                  const int    ix = std::round(std::abs(node(0) - H(0)) / dx);
                               ~~   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/jenkins/jenkins/workspace/dealii-OSX_PR-9045/source/grid/grid_generator.cc:1068:37: error: implicit conversion turns floating-point number into integer: 'double' to 'const int' [-Werror,-Wfloat-conversion]
                  const int    iy = std::round(std::abs(node(1)) / dy);
                               ~~   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Apart from the braces which are probably clear, you need to explicitly cast to an integer, like for the last line:
```
const int    iy = static_cast<int>(std::round(std::abs(node(1)) / dy));
```

kronbichler commented:
And here the errors from `clang-tidy`:
```
clang-tidy-6.0 -header-filter=/jenkins/workspace/dea/jenkins/workspace/dealii-tidy_PR-9045/source/grid/grid_generator.cc:557:31: error: the const qualified parameter 'serialnumber' is copied for each invocation; consider making it a reference [performance-unnecessary-value-param,-warnings-as-errors]
/jenkins/workspace/dealii-tidy_PR-9045/source/grid/grid_generator.cc:586:45: error: the const qualified parameter 'serialnumber' is copied for each invocation; consider making it a reference [performance-unnecessary-value-param,-warnings-as-errors]
/jenkins/workspace/dealii-tidy_PR-9045/source/grid/grid_generator.cc:613:54: error: the const qualified parameter 'serialnumber' is copied for each invocation; consider making it a reference [performance-unnecessary-value-param,-warnings-as-errors]
/jenkins/workspace/dealii-tidy_PR-9045/source/grid/grid_generator.cc:1072:37: error: C-style casts are discouraged; use static_cast [google-readability-casting,-warnings-as-errors]
```

kronbichler commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -60,6 +61,183 @@ namespace GridGenerator
+   *
+   * @ingroup GridGenerator
+   */
+  namespace Airfoil
```
I realize one thing now: Can you move this code towards the end of the file? I think it makes sense to keep the basic grids like the hypercube at the top and the more specialized ones further down.

eliasstudiert commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -60,6 +61,183 @@ namespace GridGenerator
+      /**
+       * NACA serial number defining the airfoil shape.
+       *
+       * @note Currently serial numbers with length 4 are supported.
```
Well, I've added a literature in grid_generator.cc where the 4-digit points are computed. 
(I.H. Abbott and A.E. von Doenhoff. Theory of Wing Sections: Including a Summary of Airfoil Data. New York: Dover Publications, 1959.)

I think, a good overview is given in Wikipedia, but is a link to Wikipedia okey? (https://en.wikipedia.org/wiki/NACA_airfoil)

peterrum replied:
> but is a link to Wikipedia okey?

Yes it is okay.

bangerth replied:
Yes, links are always ok. Just make sure it's linked to from the place where users want to know about it, i.e., the documentation of the function in the .h file (and not somewhere in the innards of the .cc file where a user wouldn't know to look).

eliasstudiert commented:
@kronbichler and @bangerth , I've fixed the edited your suggested modifications.
@kronbichler , I'm not sure whether I placed the code in .h at the right position and whether the image is included correctly.

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1219 @@
+
+            // copy point on lower side of airfoil
+            for (unsigned int i = nose_index; i <= tail_index; i++)
+              lower_points.emplace_back(jouk_points[i]);
```
You can probably do this more efficiently by saying
```
  lower_points.insert (lower_points.end(), jouk_points.begin()+nose_index, jouk_points.begin()+tail_index+1);
```

eliasstudiert replied:
Cool! But I don't understand why for input last, i.e. `jouk_points.begin()+tail_index+1`, the +1 is required. The last input element should be `jouk_points[tail_index]`.

kronbichler replied:
The `insert` functions, like many other C++ library functions and conventions in deal.II, work with half-open intervals, i.e., ranges where the `begin` index is included but the `end` index is excluded.

eliasstudiert replied:
Thanks!

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1219 @@
+         * ---.----|------|----|------>x
+         *     .  -1      |   .1
+         *      .         | .
+         *         .    . |
```
I love your little ASCII art pieces -- well done!

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1219 @@
+
+          // Calculate radius so that point (x=1|y=0) is enclosed - requirement
+          //  for Joukowski transform
+          double radius =
```
```suggestion
          const double radius =
```

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1219 @@
+          //  for Joukowski transform
+          double radius =
+            sqrt(center(1) * center(1) + (1 - center(0)) * (1 - center(0)));
+          double radius_test =
```
```suggestion
          const double radius_test =
```

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1219 @@
+          double radius_test =
+            sqrt(center(1) * center(1) + (1 + center(0)) * (1 + center(0)));
+          // Make sure point (x=-1|y=0) is enclosed by the circle
+          AssertThrow(
```
I suspect that you're going to get a warning about an unused variable `radius_test` in release mode. You probably want to add this:
```suggestion
          (void)radius_test;
          AssertThrow(
```

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1219 @@
+              "coordinate."));
+          // Create a full circle with radius 'radius' around Point 'center' of
+          // (number_points) equidistant points.
+          double theta = 2 * numbers::PI / number_points;
```
```suggestion
          const double theta = 2 * numbers::PI / number_points;
```

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1219 @@
+          // Calculate radius so that point (x=1|y=0) is enclosed - requirement
+          //  for Joukowski transform
+          double radius =
+            sqrt(center(1) * center(1) + (1 - center(0)) * (1 - center(0)));
```
```suggestion
            std::sqrt(center(1) * center(1) + (1 - center(0)) * (1 - center(0)));
```

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1219 @@
+          double radius =
+            sqrt(center(1) * center(1) + (1 - center(0)) * (1 - center(0)));
+          double radius_test =
+            sqrt(center(1) * center(1) + (1 + center(0)) * (1 + center(0)));
```
```suggestion
            std::sqrt(center(1) * center(1) + (1 + center(0)) * (1 + center(0)));
```

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1219 @@
+                                    const bool         is_upper)
+        {
+          // conversion string (char * ) to int
+          const unsigned int digit_0 = int(serialnumber[0]) - 48;
```
I think this might be easier to read if you wrote it like this:
```suggestion
          const unsigned int digit_0 = (serialnumber[0] - '0');
```
which I believe is the same.

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1219 @@
+          const auto scale = desired_len / input.front().distance(input.back());
+
+          for (unsigned i = 0; i < input.size(); i++)
+            output[i] = input[i] * scale;
```
I think you can write this whole function simpler by saying
```
  std::vector<Point<2>> output = input;
  const double scale = ...;
  for (auto &x : output)
    x *= scale;
  return output;
```
It's the same length in # of lines, but it's a bit easier to read what is happening.

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1219 @@
+                                           (non_equidistant_points[j + 1] -
+                                            non_equidistant_points[j]) +
+                                         non_equidistant_points[j]);
+                  i++;
```
We generally prefer the prefix form:
```suggestion
                  ++i;
```

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1219 @@
+         */
+        void make_coarse_grid(Triangulation<2> &tria) const
+        {
+          // create vector of serial triangulations for each block and temporal
```
```suggestion
          // create vector of serial triangulations for each block and temporary
```

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1219 @@
+          rotation_matrix_2[0][0] = +std::cos(+gamma);
+          rotation_matrix_2[0][1] = -std::sin(+gamma);
+          rotation_matrix_2[1][0] = +std::sin(+gamma);
+          rotation_matrix_2[1][1] = +std::cos(+gamma);
```
I *think* that `rotation_matrix_2` is just the transpose of `rotation_matrix_1`, given that `cos` is an even and `sin` an odd function. If so, just write
```
  rotation_matrix_2 = transpose(rotation_matrix_1);
```

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1219 @@
+     * approximating Joukowski or NACA (4 digit) airfoils.
+     *
+     * To specify the airfoil geometry and the mesh setup the provided
+     * inputparametrs from class AdditionalData is used. Thereby, different
```
```suggestion
     * input parameters from class AdditionalData is used. Thereby, different
```

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1219 @@
+     * size and mesh density can be created.
+     *
+     * By default, a triangulation for a NACA 2412 airfoil with chord
+     * lenght one is generated.
```
```suggestion
     * length one is generated.
```

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -28,11 +29,1219 @@
+
+    /**
+     * Initialize the given triangulation an airfoil, i.e. a Mesh of C-Type
+     * approximating Joukowski or NACA (4 digit) airfoils.
```
Is this text just a copy of the documentation you have in the .h file? If so, just delete it here -- it's in the .h file where doxygen can pick it up, and having duplication is always a prescription for letting things get out of sync.

eliasstudiert replied:
It's not a copy but it states exactly the same.  There is only this additional info in that documentation:
 `By default, a triangulation for a NACA 2412 airfoil with chord length one is generated.` 
Is this information necessary?

bangerth commented on doc/news/changes/minor/20191112EliasDejene:
```diff
@@ -0,0 +1,4 @@
+New: Add new GridGenerator tool to create a C-type mesh around a NACA or
```
By the way, I think this out to be in the "major" directory!

bangerth replied:
out -> ought.

eliasstudiert commented:
@bangerth I've edited the suggested changes and moved `20191112EliasDejene`  to major directory.

eliasstudiert commented:
@bangerth I've edited the suggested changes and moved 20191112EliasDejene to major directory.

bangerth commented:
This looks good to me now! Can you please squash/fixup the four commits into one (or, if you want, three: one for all of the code, one for the test, one for the changelog)? Then we can merge!

peterrum commented:
@bangerth Ready to merge :)

bangerth commented:
Done! Good job everyone over there, what a nice addition!

peterrum commented:
@eliasstudiert Thanks for your very good work! @blaisb Hope you like it!

eliasstudiert commented:
Thanks for helping !
